File: Algorithms/Java/Sort/InsertionSort.java
/**
 * @description: 插入排序
 * @author: WenRuo
 * @date: 2021/10/25 14:34
 */

public class InsertionSort {

    private InsertionSort() {
    }

    /**
     * 插入排序 升序
     *
     * @param arr
     * @param <E>
     */
    public static <E extends Comparable<E>> void sort2(E[] arr) {
        for (int i = 0; i < arr.length; i++) {
            for (int j = i; j - 1 >= 0 && arr[j].compareTo(arr[j - 1]) < 0; j--) {
                swap(arr, j, j - 1);
            }
        }
    }


    /**
     * 插入排序降序
     *
     * @param arr
     * @param <E>
     */
    public static <E extends Comparable<E>> void sort3(E[] arr) {
        for (int i = 0; i < arr.length; i++) {
            for (int j = i; j - 1 >= 0 && arr[j].compareTo(arr[j - 1]) > 0; j--) {
                swap(arr, j, j - 1);
            }
        }
    }

    /**
     * 插入排序优化
     *
     * @param arr
     * @param <E>
     */
    public static <E extends Comparable<E>> void sort(E[] arr) {
        for (int i = 0; i < arr.length; i++) {
            E tmp = arr[i];
            int j;
            for (j = i; j - 1 >= 0 && tmp.compareTo(arr[j - 1]) < 0; j--) {
                arr[j] = arr[j - 1];
            }
            arr[j] = tmp;
        }
    }

    private static <E> void swap(E[] arr, int i, int j) {
        E tmp = arr[i];
        arr[i] = arr[j];
        arr[j] = tmp;
    }

}

File: Algorithms/C#/Sort/InsertionSort.cs
//
// @Description:
// @Author: WenRuo
//


using System;

namespace Sort
{
    public class InsertionSort
    {
        private InsertionSort()
        {
        }

        /// <summary>
        /// 升序未优化
        /// </summary>
        /// <param name="arr"></param>
        /// <typeparam name="T"></typeparam>
        public static void Sort1<T>(T[] arr) where T : IComparable
        {
            for (int i = 0; i < arr.Length; i++)
            {
                for (int j = i; j - 1 >= 0 && arr[j].CompareTo((arr[j - 1])) < 0; j--)
                    Swap(arr, j, j - 1);
            }
        }

        /// <summary>
        /// 降序
        /// </summary>
        /// <returns></returns>
        public static void Sort2<T>(T[] arr) where T : IComparable
        {
            for (int i = 0; i < arr.Length; i++)
            {
                for (int j = i; j - 1 >= 0 && arr[j].CompareTo(arr[j - 1]) > 0; j--)
                    Swap(arr, j, j - 1);
            }
        }

        /// <summary>
        /// 升序优化
        /// </summary>
        /// <param name="arr"></param>
        /// <typeparam name="T"></typeparam>
        public static void Sort<T>(T[] arr) where T : IComparable
        {
            for (int i = 0; i < arr.Length; i++)
            {
                T tmp = arr[i];
                int j;
                for (j = i; j - 1 >= 0 && tmp.CompareTo(arr[j - 1]) < 0; j--)
                    arr[j] = arr[j - 1];
                arr[j] = tmp;
            }
        }


        private static void Swap<T>(T[] arr, int i, int j)
        {
            T tmp = arr[i];
            arr[i] = arr[j];
            arr[j] = tmp;
        }
    }
}